### PR TITLE
added strategy task

### PR DESCRIPTION
**Pull request by Nick97-git**
Added strategy task

### CODE REVIEW

**Review by Iryna-U — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -1 +1,27 @@
...
+with method `calculatePriceWithDiscount()` with one parameter `originalPrice` and 
+return type `double`. It must be located in `core.basesyntax.strategy` package.
+
+You must create its implementations (with the same names) with such discounts:
+
+| Implementation name |Discount |
+| :---: | :---: |
+| NewYearDiscountStrategy| 20 |
+| BirthdayDiscountStrategy| 33 |
+| BlackFridayDiscountStrategy| 45 |
```
Comment (boroda4436) [inline README.md:18]: You should have one class called `strategy`. And you should have one interface and multiple implementations, but they are not strategies. For example, you can have `DiscountService` or `DiscountCalculator` and have several implementations. But you still should have one class DiscountStrategy

File: README.md [hunk 2]
```diff
@@ -1 +1,27 @@
-# jv-homework-template
+# jv-strategy
+
+---
+Imagine you have сhain of stores. You need to implement discount system for people 
+in some special events of a year like New Year.
+
+In this task we'll need help of Strategy pattern.
+First of all, you need to create `DiscountStrategy` interface 
+with method `calculatePriceWithDiscount()` with one parameter `originalPrice` and
```
Comment (boroda4436) [inline README.md:9]: I guess it's better to return the amount of discount, not price minus discount

File: README.md [hunk 3]
```diff
@@ -1 +1,27 @@
...
+You must create its implementations (with the same names) with such discounts:
+
+| Implementation name |Discount |
+| :---: | :---: |
+| NewYearDiscountStrategy| 20 |
+| BirthdayDiscountStrategy| 33 |
+| BlackFridayDiscountStrategy| 45 |
+
+Each of this implementation in result of execution of method 
+`calculatePriceWithDiscount()` must return price taking into account its discount.
```
Comment (boroda4436) [inline README.md:21]: added comment above. better return amount of discount

File: README.md [hunk 4]
```diff
@@ -1 +1,27 @@
...
+Also, these implementations bust be located in `core.basesyntax.strategy.impl` package.
```
Comment (boroda4436) [inline README.md:22]:
```suggestion
Also, these implementations must be located in `core.basesyntax.strategy.impl` package.
```

File: README.md [hunk 5]
```diff
@@ -1 +1,27 @@
...
+And class context that must be called `DiscountStrategyContext`
```
Comment (boroda4436) [inline README.md:23]: What is `context` in this case? Suggest not using this class or rename it.

File: README.md [hunk 6]
```diff
@@ -1 +1,27 @@
...
+| NewYearDiscountStrategy| 20 |
+| BirthdayDiscountStrategy| 33 |
+| BlackFridayDiscountStrategy| 45 |
+
+Each of this implementation in result of execution of method 
+`calculatePriceWithDiscount()` must return price taking into account its discount.
+Also, these implementations bust be located in `core.basesyntax.strategy.impl` package.
+And class context that must be called `DiscountStrategyContext` 
+and located in `core.basesyntax` package. It must have 
+method `calculatePriceWithDiscount()` too but in this case
```
Comment (boroda4436) [inline README.md:25]: The strategy could return the specific implementation of `DiscountCalculator`.

Comment (boroda4436) [reply README.md:25]: I don't understand how the specific implementation should be chosen. We have only one method that accepts the amount. This is not enough to make a decision what implementation should be used

File: src/test/java/core/basesyntax/StrategyTest.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+public class StrategyTest {
+  private static final double DELTA = 0.0;
+  private static List<Double> discounts;
+  private static List<String> specialEvents;
+
+  @BeforeClass
+  public static void setUp() {
+    specialEvents = List.of("Birthday", "Black Friday", "New Year", "Other event");
+    discounts = List.of(33.0, 45.0, 20.0, 0.0);
```
Comment (boroda4436) [inline src/test/java/core/basesyntax/StrategyTest.java:17]: Maybe better use Map?

File: src/test/java/core/basesyntax/StrategyTest.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  }
+
+  @Test
+  public void getDiscountBySpecialEvent_IsOk() {
+    for (int i = 0; i < discounts.size(); i++) {
+      DiscountCenter discountCenter = new DiscountCenter();
```
Comment (boroda4436) [inline src/test/java/core/basesyntax/StrategyTest.java:23]: DiscountCenter is not the best name. Maybe DiscountStrategy?

File: src/test/java/core/basesyntax/StrategyTest.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+      Assert.assertEquals("Expected discount " + expectedDiscount
+              + ",but your discount " + actualDiscount,
+          expectedDiscount, actualDiscount, DELTA);
+    }
+  }
+
+  @Test
+  public void interfaceDiscountService_IsCreated() {
+    try {
+      Class<?> discountStrategyInterface = Class.forName("core.basesyntax.strategy.DiscountService");
```
Comment (boroda4436) [inline src/test/java/core/basesyntax/StrategyTest.java:35]: Which class is the direct implementation of strategy pattern? `DiscountService`, `DiscountCenter` or `BirthdayDiscountService`?

File: README.md [hunk 10]
```diff
@@ -1 +1,33 @@
...
+The last step will be creation of class called `DiscountCenter` 
+and located in `core.basesyntax` package. It must have 
+method `getDiscountBySpecialEvent(String specialEvent)` with return type `double`
+where `specialEvent` can have such values: `"Birthday"`, `"Black Friday"`, `"New Year"`.
+
+Using `specialEvent` you must write code that defines `DiscountService` implementation
+and call its `getDiscount()` method.
+
+In case if passed `specialEvent` doesn't have specific implementation
+you must return `0.0`.
```
Comment (Sofasmile) [inline README.md:33]: maybe better return DefaultDiscountService with 0.0 discount? wdyt about it?

**Review by Sofasmile — APPROVED** [review]
lgtm

File: src/test/java/core/basesyntax/StrategyTest.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+  @Test
+  public void getDiscountBySpecialEvent_IsOk() {
+    for (String specialEvent : specialEventsDiscounts.keySet()) {
+      DiscountStrategy discountStrategy = new DiscountStrategy();
+      double expectedDiscount = specialEventsDiscounts.get(specialEvent);
+      double actualDiscount = discountStrategy.getDiscountBySpecialEvent(specialEvent);
+      Assert.assertEquals("Expected discount " + expectedDiscount
+              + ",but your discount " + actualDiscount,
+          expectedDiscount, actualDiscount, DELTA);
```
Comment (boroda4436) [inline src/test/java/core/basesyntax/StrategyTest.java:29]: We should check the interface implementation we will receive as a result of method execution.
image

**Pull request merged**